グループのチャット画面への反映

akuzeee opened the pull request:
# WHAT
- チャット画面にDBから取得したグループ名やグループメンバーを表示させる
- 自分が所属していないグループのチャット画面を見たり編集できなくする
- (追加)グループメンバーが0人だった時のエラーメッセージの表示

# WHY
サービスに必須の機能なので

Commit:
Create pull request

Commit:
Update routes

Commit:
Set partial template for groups#show

Commit:
Update chat_groups#index view

Commit:
Update partial template

Commit:
Update chat_group#show view

Commit:
Update redirect after create and update

Commit:
Add redirect for unauthorized member's access

Commit:
Add error message when group_member is blank

Commit:
Fix view error in absence of group_member

Commit:
Change error_message when group_member is blank

akuzeee commented:
<img width="1280" alt="2017-02-25 13 16 29" src="https://cloud.githubusercontent.com/assets/24582005/23328089/1c088682-fb5d-11e6-96c6-6f0fcc99f1a7.png">
<img width="1280" alt="2017-02-25 13 14 28" src="https://cloud.githubusercontent.com/assets/24582005/23328090/1da13552-fb5d-11e6-9aa4-ad84d72f5de7.png">

snowhork commented on app/controllers/chat_groups_controller.rb:
```diff
@@ -31,11 +40,19 @@ def set_chat_group
   end
 
+  def set_chat_groups
+    @chat_groups = current_user.chat_groups.order('created_at DESC')
```
order(:created_at, :desc)

akuzeee replied:
order(created_at: :desc)としました

snowhork commented on app/controllers/chat_groups_controller.rb:
```diff
@@ -31,11 +40,19 @@ def set_chat_group
     params.require(:chat_group).permit(:name, user_ids: [])
   end
+
+  def redirect
```
もう少し命名工夫しよう

snowhork commented on app/controllers/chat_groups_controller.rb:
```diff
@@ -31,11 +40,19 @@ def set_chat_group
   end
+
+  def redirect
+    redirect_to :root unless GroupMember.exists?(chat_group_id: params[:id], user_id: current_user.id)
```
unless current_user.joins_in?(@chat_groups)
と書いて、
userモデルにjoins_in?(chat_group)というメソッドを作ってみましょう

snowhork commented on app/views/chat_groups/index.html.haml:
```diff
@@ -0,0 +1,19 @@
+            %p.chat-group__name #{ chat_group.name }
+            %p.chat-group__message
+              yeah!
+  = render 'show', chat_group: @chat_group if @chat_group.present?
```
showという部分テンプレートはおかしいかな。
グループ一覧のところをgroups
メッセージ一覧のところをmessages
として、

showでは両方
indexではgroupsを呼ぶという形が良さげだけどどうだろう

akuzeee replied:
グループ一覧のほうは render @chat_groups と配列を渡す形にしました

snowhork commented on config/application.rb:
```diff
@@ -14,5 +14,9 @@ class Application < Rails::Application
+
+    config.action_view.field_error_proc = Proc.new do |html_tag, instance|
+      %Q(#{html_tag}).html_safe
+    end
```
これをつけた理由を教えてください

akuzeee replied:
最初 validates_acceptance_of :user_ids によってエラーメッセージを表示させていて、
validationにはじかれたときに余計なdivが生成されるせいでチェックボックスのビューが崩れたのでその対策として入れました。
しかし validates_acceptance_of :users に変えたらビューが崩れなくなっていることに気づいたのでいま削除しました。

Commit:
Crean 'order' method by using symbol object

Commit:
Remove unnecessary description from application.rb

Commit:
Rename redirect_method for nonmember

Commit:
Clean reject_nonmember method

Commit:
Clean chat_view using partial template

snowhork commented:
ログインしてない時の挙動ってどうしてるっけ？何も設定してなければ設定しましょ

snowhork commented:
あと、グループ登録時、自分自身は必ず入っているのでは？

Commit:
Update chat_groups_controller to show error

Intended to show an error message when current user is not included in
chat_group

akuzeee commented:
> ログインしてない時の挙動ってどうしてるっけ？何も設定してなければ設定しましょ

application_controllerにbefore_action :authenticate_user!してるんですけど大丈夫でしょうか？

akuzeee commented:
> あと、グループ登録時、自分自身は必ず入っているのでは？

すみません、スマートなやり方が思いつかなかったのでごり押ししました

akuzeee commented:
<img width="1279" alt="2017-02-25 21 37 41" src="https://cloud.githubusercontent.com/assets/24582005/23331163/abffd858-fba3-11e6-80b8-60d3e93499ef.png">

snowhork commented on config/routes.rb:
```diff
@@ -1,6 +1,5 @@
-  resources :messages, only: :index
-  resources :chat_groups, only: [:new, :edit, :create, :update]
+  root       'chat_groups#index'
+  resources  :chat_groups, except: %i(destroy)
```
これ :destroy　で良さそう

snowhork commented on app/views/chat_groups/_chat_group.html.haml:
```diff
@@ -0,0 +1,5 @@
+%li.chat-group
+  = link_to chat_group_path(chat_group), class: 'chat-group__link' do
+    %p.chat-group__name #{ chat_group.name }
```
式展開の時は{}の前後で開ける必要はないかと

snowhork reviewed (commented):
ビュー側で自身は絶対チェックを入れるようになってるっけ？

Commit:
Clean routes

Commit:
Clean views
